File: backend/models/event_test.go
package models

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// frontendEventTypes mirrors frontend/src/types/event.ts (AppEvent union) exactly.
var frontendEventTypes = []string{
	"ShopCreated",
	"ShopRenamed",
	"ShopColorChanged",
	"ShopSoftDeleted",
	"TagCreated",
	"TagDeleted",
	"ItemCreated",
	"ItemUpdated",
	"ItemSoftDeleted",
	"ShopAssignedToItem",
	"ShopRemovedFromItem",
	"TagAssignedToItem",
	"TagRemovedFromItem",
	"ListCreated",
	"ListRenamed",
	"ListArchived",
	"ListUnarchived",
	"ListDeleted",
	"ListItemAdded",
	"ListItemStateChanged",
	"ListItemQuantityChanged",
	"ListItemRemoved",
	"ShopSkippedForListItem",
	"ShopSkipCleared",
	"ShoppingSessionStarted",
	"SessionItemBought",
	"SessionItemSkipped",
	"BugReported",
}

func validEvent() Event {
	return Event{
		ID:        "evt-1",
		ClientID:  "client-7",
		Lamport:   42,
		Timestamp: "2026-08-09T10:00:00Z",
		EntityID:  "shop-1",
		Type:      "ShopCreated",
		Payload:   json.RawMessage(`{"name":"Biedronka","color":"#ff0000"}`),
	}
}

func TestEventJSONRoundTripPreservesAllFields(t *testing.T) {
	orig := validEvent()

	data, err := json.Marshal(orig)
	require.NoError(t, err)

	// Wire format uses the camelCase field names the frontend sends.
	var obj map[string]json.RawMessage
	require.NoError(t, json.Unmarshal(data, &obj))
	for _, key := range []string{"id", "clientId", "lamport", "timestamp", "entityId", "type", "payload"} {
		assert.Contains(t, obj, key, "marshaled event must contain field %q", key)
	}

	var got Event
	require.NoError(t, json.Unmarshal(data, &got))
	assert.Equal(t, orig.ID, got.ID)
	assert.Equal(t, orig.ClientID, got.ClientID)
	assert.Equal(t, orig.Lamport, got.Lamport)
	assert.Equal(t, orig.Timestamp, got.Timestamp)
	assert.Equal(t, orig.EntityID, got.EntityID)
	assert.Equal(t, orig.Type, got.Type)
	assert.JSONEq(t, string(orig.Payload), string(got.Payload))
}

func TestEventUnmarshalFromClientWireShape(t *testing.T) {
	raw := `{
		"id": "e1",
		"clientId": "c1",
		"lamport": 5,
		"timestamp": "2026-08-09T10:00:00Z",
		"entityId": "item-9",
		"type": "ItemCreated",
		"payload": {"name": "Milk", "unit": "L"}
	}`
	var e Event
	require.NoError(t, json.Unmarshal([]byte(raw), &e))
	assert.Equal(t, "e1", e.ID)
	assert.Equal(t, "c1", e.ClientID)
	assert.Equal(t, 5, e.Lamport)
	assert.Equal(t, "2026-08-09T10:00:00Z", e.Timestamp)
	assert.Equal(t, "item-9", e.EntityID)
	assert.Equal(t, "ItemCreated", e.Type)
	assert.JSONEq(t, `{"name":"Milk","unit":"L"}`, string(e.Payload))
}

func TestEventTypeConstantsMatchFrontendTypeStrings(t *testing.T) {
	got := map[string]string{
		"ShopCreated":              EventShopCreated,
		"ShopRenamed":              EventShopRenamed,
		"ShopColorChanged":         EventShopColorChanged,
		"ShopSoftDeleted":          EventShopSoftDeleted,
		"TagCreated":               EventTagCreated,
		"TagDeleted":               EventTagDeleted,
		"ItemCreated":              EventItemCreated,
		"ItemUpdated":              EventItemUpdated,
		"ItemSoftDeleted":          EventItemSoftDeleted,
		"ShopAssignedToItem":       EventShopAssignedToItem,
		"ShopRemovedFromItem":      EventShopRemovedFromItem,
		"TagAssignedToItem":        EventTagAssignedToItem,
		"TagRemovedFromItem":       EventTagRemovedFromItem,
		"ListCreated":              EventListCreated,
		"ListRenamed":              EventListRenamed,
		"ListArchived":             EventListArchived,
		"ListUnarchived":           EventListUnarchived,
		"ListDeleted":              EventListDeleted,
		"ListItemAdded":            EventListItemAdded,
		"ListItemStateChanged":     EventListItemStateChanged,
		"ListItemQuantityChanged":  EventListItemQuantityChanged,
		"ListItemRemoved":          EventListItemRemoved,
		"ShopSkippedForListItem":   EventShopSkippedForListItem,
		"ShopSkipCleared":          EventShopSkipCleared,
		"ShoppingSessionStarted":   EventShoppingSessionStarted,
		"SessionItemBought":        EventSessionItemBought,
		"SessionItemSkipped":       EventSessionItemSkipped,
		"BugReported":              EventBugReported,
	}
	require.Len(t, got, len(frontendEventTypes), "one constant per frontend event type")
	for want := range got {
		assert.Equal(t, want, got[want], "constant for %q", want)
	}
}

func TestIsValidEventType(t *testing.T) {
	for _, et := range frontendEventTypes {
		assert.True(t, IsValidEventType(et), "expected %q to be a valid event type", et)
	}
	assert.False(t, IsValidEventType(""))
	assert.False(t, IsValidEventType("shopcreated"))
	assert.False(t, IsValidEventType("NoSuchEvent"))
}

func TestEventValidateOK(t *testing.T) {
	assert.NoError(t, validEvent().Validate())
}

func TestEventValidateMissingID(t *testing.T) {
	e := validEvent()
	e.ID = ""
	assert.Error(t, e.Validate())
}

func TestEventValidateMissingType(t *testing.T) {
	e := validEvent()
	e.Type = ""
	assert.Error(t, e.Validate())
}

func TestEventValidateMissingEntityID(t *testing.T) {
	e := validEvent()
	e.EntityID = ""
	assert.Error(t, e.Validate())
}

func TestEventValidateUnknownType(t *testing.T) {
	e := validEvent()
	e.Type = "NoSuchEvent"
	assert.Error(t, e.Validate())
}

func TestEventValidateNegativeLamport(t *testing.T) {
	e := validEvent()
	e.Lamport = -1
	assert.Error(t, e.Validate())
}

func TestEventValidateZeroLamportOK(t *testing.T) {
	e := validEvent()
	e.Lamport = 0
	assert.NoError(t, e.Validate())
}

func TestEventValidateInvalidPayload(t *testing.T) {
	e := validEvent()
	e.Payload = json.RawMessage(`{"name":`)
	assert.Error(t, e.Validate())
}

func TestEventValidateNilPayloadOK(t *testing.T) {
	e := validEvent()
	e.Payload = nil
	assert.NoError(t, e.Validate())
}

func TestEventValidateEmptyObjectPayloadOK(t *testing.T) {
	e := validEvent()
	e.Payload = json.RawMessage(`{}`)
	assert.NoError(t, e.Validate())
}

File: frontend/src/types/event.test.ts
import { describe, it, expect } from 'vitest'
import type { AppEvent } from './event'

function make(type: string, payload: object, entityId = 'e-1'): AppEvent {
  return {
    id: 'evt-1',
    clientId: 'client-1',
    lamport: 1,
    timestamp: '2024-01-01T00:00:00.000Z',
    entityId,
    type,
    payload,
  } as AppEvent
}

function assertNever(x: never): never {
  throw new Error(`unreachable event: ${JSON.stringify(x)}`)
}

function payloadValue(e: AppEvent): string {
  switch (e.type) {
    case 'ShopCreated': return e.payload.name
    case 'ShopRenamed': return e.payload.name
    case 'ShopColorChanged': return e.payload.color
    case 'ShopSoftDeleted': return e.payload.deletedAt
    case 'TagCreated': return e.payload.name
    case 'TagDeleted': return ''
    case 'ItemCreated': return e.payload.name
    case 'ItemUpdated': return e.payload.name ?? ''
    case 'ItemSoftDeleted': return e.payload.deletedAt
    case 'ShopAssignedToItem': return e.payload.shopId
    case 'ShopRemovedFromItem': return e.payload.shopId
    case 'TagAssignedToItem': return e.payload.tagId
    case 'TagRemovedFromItem': return e.payload.tagId
    case 'ListCreated': return e.payload.name
    case 'ListRenamed': return e.payload.name
    case 'ListArchived': return e.payload.archivedAt
    case 'ListUnarchived': return ''
    case 'ListDeleted': return e.payload.deletedAt
    case 'ListItemAdded': return e.payload.itemId
    case 'ListItemStateChanged': return e.payload.state
    case 'ListItemQuantityChanged': return String(e.payload.quantity)
    case 'ListItemRemoved': return ''
    case 'ShopSkippedForListItem': return e.payload.shopId
    case 'ShopSkipCleared': return e.payload.shopId
    case 'ShoppingSessionStarted': return e.payload.listId
    case 'SessionItemBought': return e.payload.itemId
    case 'SessionItemSkipped': return e.payload.itemId
    case 'BugReported': return e.payload.text
    default: return assertNever(e)
  }
}

describe('AppEvent union', () => {
  it('carries the full envelope on every event', () => {
    const events = [
      make('ShopCreated', { name: 'Lidl', color: '#ff0000' }),
      make('ListItemRemoved', {}),
      make('TagCreated', { name: 'dairy' }),
    ]
    for (const e of events) {
      expect(e.id).toBeDefined()
      expect(e.clientId).toBeDefined()
      expect(e.lamport).toBeGreaterThan(0)
      expect(e.timestamp).toBeDefined()
      expect(e.entityId).toBeDefined()
      expect(e.type).toBeDefined()
      expect(e.payload).toBeDefined()
    }
  })

  it('narrows the union on the type discriminator', () => {
    const created = make('ShopCreated', { name: 'Lidl', color: '#ff0000' })
    if (created.type === 'ShopCreated') {
      expect(created.payload.name).toBe('Lidl')
      expect(created.payload.color).toBe('#ff0000')
    } else {
      throw new Error('expected ShopCreated narrowing')
    }

    const renamed = make('ShopRenamed', { name: 'New' })
    if (renamed.type === 'ShopRenamed') {
      expect(renamed.payload.name).toBe('New')
    } else {
      throw new Error('expected ShopRenamed narrowing')
    }
  })

  it('handles every union member in an exhaustive switch', () => {
    const samples: AppEvent[] = [
      make('ShopCreated', { name: 'a', color: '#000' }),
      make('ShopRenamed', { name: 'b' }),
      make('ShopColorChanged', { color: '#fff' }),
      make('ShopSoftDeleted', { deletedAt: '2024-01-01T00:00:00.000Z' }),
      make('TagCreated', { name: 'dairy' }),
      make('TagDeleted', {}),
      make('ItemCreated', { name: 'Milk' }),
      make('ItemUpdated', { unit: 'kg' }),
      make('ItemSoftDeleted', { deletedAt: '2024-01-01T00:00:00.000Z' }),
      make('ShopAssignedToItem', { shopId: 's1' }),
      make('ShopRemovedFromItem', { shopId: 's1' }),
      make('TagAssignedToItem', { tagId: 't1' }),
      make('TagRemovedFromItem', { tagId: 't1' }),
      make('ListCreated', { name: 'Weekly' }),
      make('ListRenamed', { name: 'Monthly' }),
      make('ListArchived', { archivedAt: '2024-01-01T00:00:00.000Z' }),
      make('ListUnarchived', {}),
      make('ListDeleted', { deletedAt: '2024-01-01T00:00:00.000Z' }),
      make('ListItemAdded', { listId: 'l1', itemId: 'i1', state: 'active' }),
      make('ListItemStateChanged', { state: 'bought' }),
      make('ListItemQuantityChanged', { quantity: 3, unit: 'kg' }),
      make('ListItemRemoved', {}),
      make('ShopSkippedForListItem', { shopId: 's1' }),
      make('ShopSkipCleared', { shopId: 's1' }),
      make('ShoppingSessionStarted', { listId: 'l1', shopId: 's1' }),
      make('SessionItemBought', { itemId: 'i1', quantity: 2 }),
      make('SessionItemSkipped', { itemId: 'i1' }),
      make('BugReported', { text: 'crashed' }),
    ]
    expect(samples.map(payloadValue).every(v => typeof v === 'string')).toBe(true)
  })

  it('exposes payload details through type narrowing at runtime', () => {
    const li = make('ListItemAdded', { listId: 'l1', itemId: 'i1', state: 'bought', quantity: 3, unit: 'l' })
    expect(li.type).toBe('ListItemAdded')
    if (li.type === 'ListItemAdded') {
      expect(li.payload.listId).toBe('l1')
      expect(li.payload.itemId).toBe('i1')
      expect(li.payload.state).toBe('bought')
      expect(li.payload.quantity).toBe(3)
      expect(li.payload.unit).toBe('l')
    }

    const bought = make('SessionItemBought', { itemId: 'i1', quantity: 2 })
    if (bought.type === 'SessionItemBought') {
      expect(bought.payload.itemId).toBe('i1')
      expect(bought.payload.quantity).toBe(2)
    }
  })

  it('splits state changes into their own payloads', () => {
    const qty = make('ListItemQuantityChanged', { quantity: 4, unit: 'kg' })
    if (qty.type === 'ListItemQuantityChanged') {
      expect(qty.payload.quantity).toBe(4)
      expect(qty.payload.unit).toBe('kg')
    }
    const state = make('ListItemStateChanged', { state: 'active' })
    if (state.type === 'ListItemStateChanged') {
      expect(state.payload.state).toBe('active')
    }
  })
})
